--- before/opus_gemm_pipeline_a8w8_mxscale_bpreshuffle_8wave_192x256_gfx950.cuh
+++ after/opus_gemm_pipeline_a8w8_mxscale_bpreshuffle_8wave_192x256_gfx950.cuh
@@ -182,12 +182,8 @@
     const int loops = kargs.k / T::B_K;
 
     // Matrix and scale global-memory views.
-    const auto b_address = reinterpret_cast<u64_t>(kargs.ptr_b) +
-                           static_cast<u64_t>(col) * kargs.stride_b;
-    const u64_t b_lo = static_cast<unsigned>(__builtin_amdgcn_readfirstlane(static_cast<unsigned>(b_address)));
-    const u64_t b_hi = static_cast<unsigned>(__builtin_amdgcn_readfirstlane(static_cast<unsigned>(b_address >> 32)));
     auto g_a = make_gmem(reinterpret_cast<const D_A*>(kargs.ptr_a) + row * kargs.stride_a, static_cast<unsigned>((kargs.m - row) * kargs.stride_a));
-    auto g_b = make_gmem(reinterpret_cast<const D_B*>(b_lo | (b_hi << 32)));
+    auto g_b = make_gmem(reinterpret_cast<const D_B*>(kargs.ptr_b) + col * kargs.stride_b);
     auto g_c = make_gmem(reinterpret_cast<D_C*>(kargs.ptr_c) + row * kargs.stride_c + col, static_cast<unsigned>(((kargs.m - row) * kargs.stride_c - col) * sizeof(D_C)));
     auto g_sfa = make_gmem(reinterpret_cast<const D_SF*>(kargs.ptr_sfa) + row);
     auto g_sfb = make_gmem(reinterpret_cast<const D_SF*>(kargs.ptr_sfb) + (col / T::GROUP_N) * kargs.stride_sfb);
